Feature antikler bjs2 65039

antikler opened the pull request.

antikler committed:
BJS2-65039 Become Mentor.
release all method of task:
1) getMentees(long userId)
2) getMentors(long userId)
3) deleteMentee(long menteeId, long mentorId)
4) deleteMentor(long menteeId, long mentorId)

Commit:
BJS2-65039 Become Mentor: added RequestMapping

DmitryRogulkin commented on src/main/java/school/faang/user_service/mapper/MentorshipMapper.java:
```diff
@@ -0,0 +1,54 @@
+@Mapper(componentModel = "spring")
+public interface MentorshipMapper {
+
+    @Mapping(target = "userId", source = "entity.id")
```
Бросильось в глаза оргомное количество @Mapping(target = ..., ignore = true), что может говорить о необходимости создания более специализированного DTO. На сколько я видел задания у тебя там явно требования по dto не указаны. Не уверен, что надо перечислять все поля. Например включая password, который как бы не передают по сети. Об этом говорил Влад в лекции.

DmitryRogulkin commented on src/main/java/school/faang/user_service/service/MentorshipService.java:
```diff
@@ -0,0 +1,55 @@
+    private final MentorshipRepository mentorshipRepository;
+    private final MentorshipMapper mentorshipMapper;
+
+    public List<MentorshipDto> getMentees(long userId) {
```
Метод возвращает список менторов в виде DTO, но в задании сказано, что он должен возвращать список пользователей-менти. Как будто тут надо List<UserDto>

DmitryRogulkin replied:
Не допечаталось List<UserDto>

DmitryRogulkin replied:
UserDTO

Commit:
BJS2-65039 Become Mentor: new realisation task

Commit:
BJS2-65039 Become Mentor: bug fixes

Soddik commented on src/main/java/school/faang/user_service/controller/mentorship/MentorshipController.java:
```diff
@@ -0,0 +1,39 @@
+    private final MentorshipService mentorshipService;
+
+    @GetMapping("/mentors/{userId}/mentees")
+    public List<MentorshipDto> getMentees(@PathVariable long userId) {
```
для аннотации лучше всегда указывать имя @PathVariable("variable_name")
и нэйминг, лучше будет /mentors/{mentorId}/mentees

antikler replied:
Исправил.

Soddik commented on src/main/java/school/faang/user_service/controller/mentorship/MentorshipController.java:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    @GetMapping("/mentees/{userId}/mentors")
+    public List<MentorshipDto> getMentors(@PathVariable long userId) {
```
для аннотации лучше всегда указывать имя @PathVariable("variable_name")
и нэйминг, лучше будет /mentees/{menteeId}/mentors

antikler replied:
Исправил.

Soddik commented on src/main/java/school/faang/user_service/controller/mentorship/MentorshipController.java:
```diff
@@ -0,0 +1,39 @@
+        return mentorshipService.getMentors(userId);
+    }
+
+    @DeleteMapping("/mentors/{mentorId}/mentees/{menteeId}")
```
для аннотации лучше всегда указывать имя @PathVariable("variable_name")

antikler replied:
Исправил.

Soddik commented on src/main/java/school/faang/user_service/controller/mentorship/MentorshipController.java:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    @DeleteMapping("/mentees/{menteeId}/mentors/{mentorId}")
+    public void deleteMentor(@PathVariable long menteeId, @PathVariable long mentorId) {
```
для аннотации лучше всегда указывать имя @PathVariable("variable_name")

antikler replied:
Исправил.

Soddik commented on src/main/java/school/faang/user_service/controller/mentorship/MentorshipController.java:
```diff
@@ -0,0 +1,39 @@
+@RestController
+@RequestMapping("/mentorship")
+@RequiredArgsConstructor
+public class MentorshipController {
```
во всем контроллере нет обертки ResponseEntity с статусом ответа

antikler replied:
Сделал через аннотацию для каждого метода. Или нужно обязательно именно через
return ResponseEntity.ok("Some message")?

Soddik replied:
лучше ResponseEntity с возможностью изменять тело и статусы динамически

antikler replied:
Исправил на ResponseEntity

Soddik commented on src/main/java/school/faang/user_service/service/MentorshipService.java:
```diff
@@ -0,0 +1,55 @@
+
+    public void deleteMentee(long menteeId, long mentorId) {
+        User mentor = userRepository.findById(mentorId)
+                .orElseThrow(() -> new IllegalStateException("Mentor not found."));
```
точно IllegalStateException ?

antikler replied:
Исправил на EntityNotFoundException, либо сделать свой обработчик ошибок (видел у Романа реализацию, поэтому свой не стал писать.)

Soddik commented on src/main/java/school/faang/user_service/service/MentorshipService.java:
```diff
@@ -0,0 +1,55 @@
+        User mentor = userRepository.findById(mentorId)
+                .orElseThrow(() -> new IllegalStateException("Mentor not found."));
+        User mentee = userRepository.findById(menteeId)
+                .orElseThrow(() -> new IllegalStateException("Mentee not found."));
```
точно IllegalStateException ?

antikler replied:
Исправил на EntityNotFoundException

Soddik commented on src/main/java/school/faang/user_service/service/MentorshipService.java:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    public void deleteMentor(long menteeId, long mentorId) {
+        deleteMentee(menteeId, mentorId);
```
можно было вынести в private метод и вызывать отдельно в каждом из public методов, сейчас получатся связь между public методами

antikler replied:
Исправил.

Soddik commented on src/main/java/school/faang/user_service/service/MentorshipService.java:
```diff
@@ -0,0 +1,55 @@
+                .orElseGet(List::of);
+    }
+
+    public void deleteMentee(long menteeId, long mentorId) {
```
нужен ли @Transactional ?

antikler replied:
Да, нужен, т.к. изменяю 2 сущности.
Исправил.

Soddik commented on src/main/java/school/faang/user_service/service/MentorshipService.java:
```diff
@@ -0,0 +1,55 @@
+        userRepository.save(mentee);
+    }
+
+    public void deleteMentor(long menteeId, long mentorId) {
```
нужен ли @Transactional ?

antikler replied:
Да, нужен, т.к. изменяю 2 сущности.
Исправил.

Soddik commented on src/main/java/school/faang/user_service/mapper/MentorshipMapper.java:
```diff
@@ -0,0 +1,23 @@
+    @Mapping(target = "username", source = "dto.userName")
+    User toEntity(MentorshipDto dto);
+
+    List<MentorshipDto> toDtos(List<User> entities);
```
зачем если не используется?

antikler replied:
Делал по аналогии с другими мапперами в проекте. В первых попытках реализации задачи пробовал вернуть списки Dto, но потом от реализации отказался, а метод остался.
Но теперь по рекомендации использую.

Soddik commented on src/main/java/school/faang/user_service/repository/UserRepository.java:
```diff
@@ -5,6 +5,7 @@
 import java.util.List;
+import java.util.Optional;
 import java.util.stream.Stream;
 
```
а здесь @Repository не поставил

antikler replied:
Spring умный, а я не внимательный.
Исправил.

Soddik commented on src/main/java/school/faang/user_service/service/MentorshipService.java:
```diff
@@ -0,0 +1,55 @@
+
+    public List<MentorshipDto> getMentees(long userId) {
+        return userRepository.findById(userId)
+                .map(User::getMentees)
```
можно же так
```java
public List<MentorshipDto> getMentees(long userId) {
        return userRepository.findById(userId)
            .map(User::getMentees)
            .map(mentorshipMapper::toDtos)
            .orElse(List.of());
    }
```

antikler replied:
Можно, но переделал согласно совета ниже. Самого смущало решение, ведь код однообразный, но так использовать функцию я не догадался.
Благодарю!

Soddik commented on src/main/java/school/faang/user_service/service/MentorshipService.java:
```diff
@@ -0,0 +1,55 @@
+                .collect(Collectors.toList());
+    }
+
+    public List<MentorshipDto> getMentors(long userId) {
```
а тут так:
```java
public List<MentorshipDto> getMentors(long userId) {
        return userRepository.findById(userId)
            .map(User::getMentors)
            .map(mentorshipMapper::toDtos)
            .orElse(List.of());
    }
```

antikler replied:
Исправил

Soddik commented on src/main/java/school/faang/user_service/service/MentorshipService.java:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    public List<MentorshipDto> getMentors(long userId) {
+        return userRepository.findById(userId)
```
а вообще оба метода можно свести к:
```java
public List<MentorshipDto> getMentees(long userId) {
        return getUsersById(userId, User::getMentees);
    }

    public List<MentorshipDto> getMentors(long userId) {
        return getUsersById(userId, User::getMentors);
    }

    private List<MentorshipDto> getUsersById(long userId, Function<User, List<User>> mappingFunction) {
        return userRepository.findById(userId)
            .map(mappingFunction)
            .map(mentorshipMapper::toDtos)
            .orElse(List.of());
    }
```

antikler replied:
Принял на вооружение.

antikler reviewed (commented):
Исправил.

Commit:
BJS2-65039 Become Mentor: bug fixes after Alexey's code-review

antikler committed:
BJS2-65039 Become Mentor. Refactor code:
 1) ResponseStatus vs ResponseEntity in MentorshipController.
 2) Transfer annotation @Transactional from private method to public method in MentorshipService.
